lib/rs/src/mock/mod.rs

```diff
@@ -0,0 +1,255 @@
+        read!(self, Message(Begin((name, type_, id))), (name, type_, id))
+    }
+
+    fn read_message_end<T: Transport>(&mut self, _: &mut T) -> Result<()> {
+        read!(self, Message(End))
+    }
+
+    fn read_struct_begin<T: Transport>(&mut self, _: &mut T) -> Result<String> {
+        read!(self, Struct(Begin(name)), name)
+    }
+
+    fn read_struct_end<T: Transport>(&mut self, _: &mut T) -> Result<()> {
+        read!(self, Struct(End))
+    }
+
+    fn read_field_begin<T: Transport>(&mut self, _: &mut T) -> Result<(String, Type, i16)> {
+        read!(self, Field(Begin((name, type_, id))), (name, type_, id))
+    }
+
+    fn read_field_end<T: Transport>(&mut self, _: &mut T) -> Result<()> {
+        read!(self, Field(End))
+    }
+
+    fn read_map_begin<T: Transport>(&mut self, _: &mut T) -> Result<(Type, Type, i32)> {
+        read!(self, Map(Begin((keyt, valuet, len))), (keyt, valuet, len as i32))
+    }
+
+    fn read_map_end<T: Transport>(&mut self, _: &mut T) -> Result<()> {
+        read!(self, Map(End))
+    }
+
+    fn read_list_begin<T: Transport>(&mut self, _: &mut T) -> Result<(Type, i32)> {
+        read!(self, List(Begin((ty, len))), (ty, len as i32))
+    }
+    fn read_list_end<T: Transport>(&mut self, _: &mut T) -> Result<()> {
+         read!(self, List(End))
+    }
+
+    fn read_set_begin<T: Transport>(&mut self, _: &mut T) -> Result<(Type, i32)> {
+        read!(self, Set(Begin((ty, len))), (ty, len as i32))
+    }
+
+    fn read_set_end<T: Transport>(&mut self, _: &mut T) -> Result<()> {
+         read!(self, Set(End))
+    }
+
+    fn read_bool<T: Transport>(&mut self, _: &mut T) -> Result<bool> { read!(self, Prim(Bool(val)), val) }
+    fn read_byte<T: Transport>(&mut self, _: &mut T) -> Result<i8> { read!(self, Prim(Byte(val)), val) }
+    fn read_i16<T: Transport>(&mut self, _: &mut T) -> Result<i16> { read!(self, Prim(I16(val)), val) }
+    fn read_i32<T: Transport>(&mut self, _: &mut T) -> Result<i32> { read!(self, Prim(I32(val)), val) }
+    fn read_i64<T: Transport>(&mut self, _: &mut T) -> Result<i64> { read!(self, Prim(I64(val)), val) }
+    fn read_double<T: Transport>(&mut self, _: &mut T) -> Result<f64> { read!(self, Prim(Double(val)), val) }
+    fn read_string<T: Transport>(&mut self, _: &mut T) -> Result<String> { read!(self, Prim(PString(string)), string) }
+    fn read_binary<T: Transport>(&mut self, _: &mut T) -> Result<Vec<u8>> { read!(self, Prim(Binary(val)), val) }
+
+    fn skip<T: Transport>(&mut self, _: &mut T, _: Type) -> Result<()> {
+        // TODO: Implement *checked* skipping
+        if self.log.len() != 0 { self.log.pop(); }
+        Ok(())
+    }
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum ProtocolAction {
+    Message(Action<(String, MessageType, i32)>),
+    Struct(Action<String>),
+    Field(Action<(String, Type, i16)>),
+    Map(Action<(Type, Type, usize)>),
+    List(Action<(Type, usize)>),
+    Set(Action<(Type, usize)>),
+    Prim(Primitive)
```

Comment:
nit: I've been getting in the habit of adding trailing commas to struct and enum declarations, I think we should do that here and in the other struct/enum declarations.